# 作业4 空间查询处理和优化

**作业目的：**理解(空间)查询处理与优化流程，掌握关系代数优化和基于代价的查询规划，掌握空间填充曲线在常用空间查询下的代价，掌握PostgreSQL的查询规划和GiST索引创建与使用，理解索引在基于代价的查询规划中的作用。

**注意事项：**
* SQL语句的错误输出为乱码时，修改SET client_encoding = 'GBK';或SET client_encoding = 'UTF-8';，重新连接数据库
* Jupyter Notebook对SQL语句的错误提示较弱，可以先在pgAdmin 4上执行，查看详细的错误信息
* 作业4总分50分，作业考察的题目后面标了具体分数，可以相互讨论思路，作业抄袭或雷同都要扣分
* **作业4\_学号\_姓名.ipynb**替换其中的学号和姓名，包含执行结果，压缩为__作业4\_学号\_姓名.rar/zip__，**不要包含数据文件**，提交到学在浙大，作业4截止日期**2020.4.26**

### 1. 空间计算（2分）

2016年Communications of The ACM上发表了一篇[Spatial Computing](http://www.cad.zju.edu.cn/home/ybtao/sdb/resources/CACM2016.pdf)论文，中文翻译[空间计算](http://www.cad.zju.edu.cn/home/ybtao/sdb/resources/CCCF2016.pdf)，其主要观点如下：
* Starting with the public availability of GPS, spatial computing has enriched our lives through location-based services (such as Google Maps, Uber, geotagging, and geotargeted, including Amber, alerts).
* It has also advanced computer science through ideas like spatial databases (such as R-tree and OGIS simple features library), spatial statistics (such as point process theory and Kriging), and spatial data mining (such as robust hotspot detection).
* Future potentially transformative opportunities include ubiquitous indoor location-based services, the location-aware Internet of Physical Things, continuous global monitoring, visualization, forecast, alerts, and warnings to address societal challenges like climate change and how to provide adequate food, energy, and water.

阅读该论文，根据文中内容回答以下问题。

1.1 空间数据库的哪些方面在已有的空间计算中发挥了重要作用？（1分）

1.2 从近期和长远来看，空间计算将如何促进空间数据库和空间统计的发展？（1分）

### 2. 代价估计与索引（14分）

健身俱乐部设计了如下数据库，用来记录俱乐部会员：<br/>
&ensp;&ensp;&ensp; Gym (<u>gid</u>, name, city)<br/>
&ensp;&ensp;&ensp; Member (<u>mid</u>, name, is_student, birthdate, city)<br/>
&ensp;&ensp;&ensp; Visits (<u>timestamp</u>, <u>mid</u> References Member, gid References Gym)<br/>

Member和Visits在数据库中的统计信息如下：<br/>
T(Member) = 500, B(Member) = 100, V(Member, city) = 10, V(Member, is_student) = 2<br/>
T(Visits) = 5000, B(Visits) = 400, V(Visits, mid) = 500<br/>
其中T表示关系的记录数目，B表示关系的磁盘页或Block数目，V表示属性不同取值的个数。

3.1 对于查询<br/>
&ensp;&ensp;&ensp; Select name <br/>
&ensp;&ensp;&ensp; From Member <br/>
&ensp;&ensp;&ensp; Where city = '杭州' and is_student = True;<br/>
基于下列条件估计磁盘I/O cost，即Block的读取块数。对于$\sigma_{a=?}(R)$查询，返回的记录数目为T(R) * 1 / V(R,a) (5分)

2.1.1 没有索引

2.1.2 Member的city属性上有非聚集索引

2.1.3 Member的city属性上有聚集索引

2.1.4 Member的is_student属性上有非聚集索引

2.1.5 Member的(city, is_student)属性上有非聚集索引

3.2 对于查询<br/>
&ensp;&ensp;&ensp; Select name, count(\*) <br/>
&ensp;&ensp;&ensp; From Member M, Visits V <br/>
&ensp;&ensp;&ensp; Where M.mid = V.mid and city = '杭州' and is_student = True <br/>
&ensp;&ensp;&ensp; Group By name;<br/>
考虑查询中的Join部分，可以使用nested loop, merge join, hash join等算法。假设内存足够大，没有临时表存储到磁盘，基于以下等价的关系代数查询树，估计磁盘I/O cost。对于$R\Join S$查询，nested loop的I/O cost为B(R) + B(R) \* B(S)或B(S) + B(R) \* B(S)。(5分)
<img src = "querytree.png"/>

2.2.1 基于左边查询树(先Join后Select)，没有索引，使用nested loop的最小I/O cost

2.2.2 基于左边查询树(先Join后Select)，Visits的mid属性上有非聚集索引，使用nested loop with index的最小I/O cost

2.2.3 基于左边查询树(先Join后Select)，Visits的mid属性上有聚集索引，使用nested loop with index的最小I/O cost

2.2.4 基于右边查询树(先Select后Join)，Member的(city,is_student)属性上有非聚集索引，使用nested loop的的I/O cost

2.2.5 基于右边查询树(先Select后Join)，Member的(city,is_student)属性上有非聚集索引，Visits的mid属性上有非聚集索引，使用nested loop的的I/O cost

2.3 你构建了索引选择工具来度量某个索引对常用查询的性能提升(越高越好)。索引1(city, is_student)带来的性能提升是10，索引2(city, is_student, birthday)带来的性能提升是12，索引3(city, birthday)带来的性能提升是7。假设你只能保留2个索引，你会选择哪两个索引，理由是什么？(2分)

2.4 健身俱乐部发现会员签到非常慢，邀请你来优化他们的数据库。在分析数据库后，你发现属性上创建了很多索引，包括Visits上的mid聚集索引。为什么mid聚集索引会使用户签到变慢，即插入一个新的Visit变慢？你需要什么信息来决定保留哪些索引和删除哪些索引？(2分)

2.5 (附加题) 你和你的朋友同时到达杭州紫金港健身俱乐部，一起签到时，发生了如下错误：

    Error: Duplicate entry '1586224200' for key 'pk_visits'
    
你觉得DBA在创建关系时做错了什么，导致了这个错误？如何进行修正这个错误？(2分)

### 3. 空间填充曲线 (10分)

15个点存储在空间数据库中，如下图标识的A-O点，假设每个数据页最多存储两个数据点，Point Query和Nearest Neighbor Query为红色查询点，Range Query为黑色查询框。
<img src = "curve.png"/>

3.1 假设这些点使用heap file存储，即乱序存储，以下查询需要访问多少数据页？(3分)

3.2 构建4x4的Hilbert Curve，数据点按照Hilbert value的顺序存储，使用()表示一个数据页，给出数据点在数据库中的存储顺序。(2分)

3.3 基于Hilbert Curve存储的数据点，以下查询需要访问最少多少数据页？假设内存足够大，不需要重复读取数据页，未使用B+Tree对Hilbert Value构建索引。(5分)

### 4. 关系代数表达式优化（6分）

通过pgAdmin 4在PostgreSQL数据库中创建hw4数据库，增加postgis扩展，并连接该数据库

In [2]:
%load_ext sql
import time
import random

# To help render markdown
from IPython.core.display import display, HTML
from markdown import markdown
def render_markdown_raw(m): return display(HTML(markdown(m))) # must be last element of cell.
def render_markdown(m): return render_markdown_raw(m.toMD())
def cost_markdown(q): 
    q.reset_count()
    get_result(q) # run the counters
    return display(HTML(markdown("Total Reads: {0}\n\n".format(q.total_count()) + q.toCount(0))))

# import the relational algbera operators
from relation_algebra import Select, Project, Union, NJoin, CrossProduct, BaseRelation
from relation_algebra import get_result, compare_results

In [3]:
%%sql postgresql://postgres:postgres@localhost:5432/hw4

SET statement_timeout = 0;
SET lock_timeout = 0;
SET client_encoding = 'utf-8';
SET standard_conforming_strings = on;
SET check_function_bodies = false;

Done.
Done.
Done.
Done.
Done.


[]

关系代数表达式优化参考练习7。创建表R, S, T，并插入数据

In [5]:
%%sql
drop table if exists R; create table R(a int, b int);
drop table if exists S; create table S(b int, c int);
drop table if exists T; create table T(c int, d int);

 * postgresql://postgres:***@localhost:5432/hw4
Done.
Done.
Done.
Done.
Done.
Done.


[]

In [6]:
v1 = "";
v2 = "";
for b in range(0,10,1):
    for a in range(0,20,2):
        t1 = " (%d, %d)," % (a, b)
        t2 = " (%d, %d)," % (b, a)
        v1 = v1 + t1
        v2 = v2 + t2
r = "insert into R values" + v1[:-1] + ";"
s = "insert into S values" + v2[:-1] + ";"
t = "insert into T values" + v2[:-1] + ";"
%sql delete from R; $r
%sql delete from S; $s
%sql delete from T; $t

 * postgresql://postgres:***@localhost:5432/hw4
0 rows affected.
100 rows affected.
 * postgresql://postgres:***@localhost:5432/hw4
0 rows affected.
100 rows affected.
 * postgresql://postgres:***@localhost:5432/hw4
0 rows affected.
100 rows affected.


[]

回顾关系代数表达式基本形式

In [8]:
r = %sql SELECT * FROM R;
R = BaseRelation(r, name="R")
s = %sql SELECT * FROM S;
S = BaseRelation(s, name="S")
t = %sql SELECT * FROM T;
T = BaseRelation(t, name="T")

x = Project(["b"], NJoin(R,S))
render_markdown(x)
print(get_result(x))

 * postgresql://postgres:***@localhost:5432/hw4
100 rows affected.
 * postgresql://postgres:***@localhost:5432/hw4
100 rows affected.
 * postgresql://postgres:***@localhost:5432/hw4
100 rows affected.


[(0,), (1,), (2,), (8,), (3,), (9,), (4,), (5,), (6,), (7,)]


熟悉cost_markdown函数

In [9]:
cost_markdown(x)

在关系数据库系统中，cost主要是I/O cost，即数据页或数据块读取次数（**注意和空间数据库系统的差异**）。在计算cost时，做了以下假设：1. 存储系统没有cache，无论是buffer management还是磁盘上的cache；2. 自然连接实现方式，是基于什么算法？通过构造等价的关系代数表达式，优化以下查询。
#### 4.1

In [10]:
x = Select("a", 2, Project(["a","c"], NJoin(R,S)))
render_markdown(x)
print(get_result(x))
cost_markdown(x)

[(2, 6), (2, 18), (2, 12), (2, 2), (2, 8), (2, 14), (2, 4), (2, 16), (2, 10), (2, 0)]


In [11]:
y = Project(["a","c"], NJoin(Select("a", 2, R), S))
render_markdown(y)
print(get_result(y))
print(compare_results(x,y))
cost_markdown(y)

[(2, 16), (2, 6), (2, 18), (2, 8), (2, 10), (2, 0), (2, 12), (2, 2), (2, 14), (2, 4)]
True


#### 4.2（2分）

In [12]:
x = Select("c", 0, Project(["a","c"], Select("b", 0, NJoin(NJoin(R, S), T))))
render_markdown(x)
print(get_result(x))
cost_markdown(x)

[(18, 0), (8, 0), (14, 0), (4, 0), (10, 0), (0, 0), (6, 0), (16, 0), (12, 0), (2, 0)]


In [25]:
y =  Project(["a","c"], NJoin(NJoin(Select("c", 0, Select("b", 0, S)),Select("c", 0, T) ), Select("b", 0, R)))
render_markdown(y)
print(get_result(y))
print(compare_results(x,y))
cost_markdown(y)

[(18, 0), (0, 0), (12, 0), (8, 0), (6, 0), (2, 0), (16, 0), (14, 0), (10, 0), (4, 0)]
True


#### 4.3（4分）

In [57]:
x = Select("c", 0, Project(["c"], Select("d", 2, Select("a", 4, NJoin(R, NJoin(S,T))))))
render_markdown(x)
print(get_result(x))
cost_markdown(x)

[(0,)]


In [59]:
y =  Project(["c"],NJoin(Select("c", 0,Select("d", 2, T)), NJoin(Select("c", 0, S), Select("a", 4, R ))))
render_markdown(y)
print(get_result(y))
print(compare_results(x,y))
cost_markdown(y)

[(0,)]
True


In [33]:
%%sql 
explain analyze
select p.c from
(select S.c from R, S, T where S.c = T.c and S.b = R.b and R.a = 4 and T.d = 2) 
p where p.c = 0;

 * postgresql://postgres:***@localhost:5432/hw4
16 rows affected.


[('Nested Loop  (cost=2.38..7.47 rows=10 width=4) (actual time=0.036..0.046 rows=10 loops=1)',),
 ('  ->  Seq Scan on t  (cost=0.00..2.50 rows=1 width=4) (actual time=0.012..0.016 rows=1 loops=1)',),
 ('        Filter: ((c = 0) AND (d = 2))',),
 ('        Rows Removed by Filter: 99',),
 ('  ->  Hash Join  (cost=2.38..4.88 rows=10 width=4) (actual time=0.024..0.029 rows=10 loops=1)',),
 ('        Hash Cond: (r.b = s.b)',),
 ('        ->  Seq Scan on r  (cost=0.00..2.25 rows=10 width=4) (actual time=0.006..0.010 rows=10 loops=1)',),
 ('              Filter: (a = 4)',),
 ('              Rows Removed by Filter: 90',),
 ('        ->  Hash  (cost=2.25..2.25 rows=10 width=8) (actual time=0.013..0.013 rows=10 loops=1)',),
 ('              Buckets: 1024  Batches: 1  Memory Usage: 9kB',),
 ('              ->  Seq Scan on s  (cost=0.00..2.25 rows=10 width=8) (actual time=0.005..0.010 rows=10 loops=1)',),
 ('                    Filter: (c = 0)',),
 ('                    Rows Removed by Filter: 90',),
 ('Planning Time: 0.100 ms',),
 ('Execution Time: 0.064 ms',)]

In [75]:
%%sql 
drop index if exists ri;
create index ri on R(a asc); 

 * postgresql://postgres:***@localhost:5432/hw4
Done.
Done.


[]

In [76]:
%%sql 
explain analyze
select p.c from
(select S.c from R, S, T where S.c = T.c and S.b = R.b and R.a = 4 and T.d = 2) 
p where p.c = 0;

 * postgresql://postgres:***@localhost:5432/hw4
16 rows affected.


[('Nested Loop  (cost=2.38..7.47 rows=10 width=4) (actual time=0.032..0.042 rows=10 loops=1)',),
 ('  ->  Seq Scan on t  (cost=0.00..2.50 rows=1 width=4) (actual time=0.009..0.013 rows=1 loops=1)',),
 ('        Filter: ((c = 0) AND (d = 2))',),
 ('        Rows Removed by Filter: 99',),
 ('  ->  Hash Join  (cost=2.38..4.88 rows=10 width=4) (actual time=0.022..0.028 rows=10 loops=1)',),
 ('        Hash Cond: (r.b = s.b)',),
 ('        ->  Seq Scan on r  (cost=0.00..2.25 rows=10 width=4) (actual time=0.006..0.010 rows=10 loops=1)',),
 ('              Filter: (a = 4)',),
 ('              Rows Removed by Filter: 90',),
 ('        ->  Hash  (cost=2.25..2.25 rows=10 width=8) (actual time=0.013..0.013 rows=10 loops=1)',),
 ('              Buckets: 1024  Batches: 1  Memory Usage: 9kB',),
 ('              ->  Seq Scan on s  (cost=0.00..2.25 rows=10 width=8) (actual time=0.005..0.010 rows=10 loops=1)',),
 ('                    Filter: (c = 0)',),
 ('                    Rows Removed by Filter: 90',),
 ('Planning Time: 1.498 ms',),
 ('Execution Time: 0.058 ms',)]

In [42]:
# 优化过后的关系代数表达式如下：
y =  Project(["c"],NJoin(Select("c", 0,Select("d", 2, T)), NJoin(Select("c", 0, S), Select("a", 4, R ))))
render_markdown(y)

### 5. 空间索引（18分）

非空间索引相关练习参考课堂练习6。

Natural Earth网站下载[Urban Areas](http://www.naturalearthdata.com/downloads/10m-cultural-vectors/) (12.49 MB version 4.0.0)，[Rivers+lake centerlines](http://www.naturalearthdata.com/downloads/10m-physical-vectors/) (1.73 MB version 4.1.0)，和[Populated Places](http://www.naturalearthdata.com/downloads/10m-cultural-vectors/) (2.68 MB version 4.1.0)数据。

在PostgreSQL中导入上述三个shapefile文件，注意导入时**缺省选项是为空间数据创建空间索引**，需要取消这些创建索引选项，手动为Urban_Areas、Rivers和Cities创建空间索引，需要手动修改导入后的关系名。

<img src = "shapefile.png"/>

构造以下空间查询，不能使用with语句。首先在无空间索引下查询，然后创建空间索引，再在有空间索引下查询。

#### 5.1 查询每条河流穿越城市区域的次数，查询结果模式为(rivers.name, num)，按次数降序排列（3分）

In [112]:
%%sql
explain analyze
select r.name,count(*) num from ne_10m_rivers_lake_centerlines r, ne_10m_urban_areas u 
where r.featurecla = 'River' and ST_Crosses(r.geom,u.geom) group by r.name order by num desc;

 * postgresql://postgres:***@localhost:5432/hw4
23 rows affected.


[('Sort  (cost=212488009.42..212488011.86 rows=974 width=16) (actual time=38062.753..38062.784 rows=490 loops=1)',),
 ('  Sort Key: (count(*)) DESC',),
 ('  Sort Method: quicksort  Memory: 52kB',),
 ('  ->  Finalize GroupAggregate  (cost=212487129.56..212487961.07 rows=974 width=16) (actual time=38061.587..38062.636 rows=490 loops=1)',),
 ('        Group Key: r.name',),
 ('        ->  Gather Merge  (cost=212487129.56..212487946.46 rows=974 width=16) (actual time=38061.576..38067.433 rows=743 loops=1)',),
 ('              Workers Planned: 1',),
 ('              Workers Launched: 1',),
 ('              ->  Partial GroupAggregate  (cost=212486129.55..212486836.88 rows=974 width=16) (actual time=38028.951..38029.279 rows=372 loops=2)',),
 ('                    Group Key: r.name',),
 ('                    ->  Sort  (cost=212486129.55..212486362.08 rows=93012 width=8) (actual time=38028.939..38029.004 rows=934 loops=2)',),
 ('                          Sort Key: r.name',),
 ('                          Sort Method: quicksort  Memory: 67kB',),
 ('                          Worker 0:  Sort Method: quicksort  Memory: 70kB',),
 ('                          ->  Nested Loop  (cost=0.00..212478453.67 rows=93012 width=8) (actual time=77.713..38025.971 rows=934 loops=2)',),
 ('                                Join Filter: st_crosses(r.geom, u.geom)',),
 ('                                Rows Removed by Join Filter: 7137745',),
 ('                                ->  Parallel Seq Scan on ne_10m_urban_areas u  (cost=0.00..2333.87 rows=6987 width=1601) (actual time=0.007..6.605 rows=5939 loops=2)',),
 ('                                ->  Seq Scan on ne_10m_rivers_lake_centerlines r  (cost=0.00..348.19 rows=1202 width=2874) (actual time=0.001..0.297 rows=1202 loops=11878)',),
 ("                                      Filter: ((featurecla)::text = 'River'::text)",),
 ('                                      Rows Removed by Filter: 253',),
 ('Planning Time: 0.240 ms',),
 ('Execution Time: 38067.964 ms',)]

基于上述查询规划回答以下问题

#### 5.2 查询在亚马逊河流10个单位距离内的所有城市，查询结果模式为(cities.gid, cities.name) (使用ST_Distance实现)（3分）

In [20]:
%%sql 
explain analyze
select distinct p.gid, p.name from ne_10m_rivers_lake_centerlines r, ne_10m_populated_places p 
where r.featurecla = 'River' and r.name_zh like '%亚马逊%' and ST_Distance(r.geom,p.geom) < 10;

 * postgresql://postgres:***@localhost:5432/hw4
13 rows affected.


[('Unique  (cost=184968.83..184987.19 rows=2448 width=13) (actual time=149.310..149.349 rows=273 loops=1)',),
 ('  ->  Sort  (cost=184968.83..184974.95 rows=2448 width=13) (actual time=149.308..149.318 rows=273 loops=1)',),
 ('        Sort Key: p.gid, p.name',),
 ('        Sort Method: quicksort  Memory: 39kB',),
 ('        ->  Nested Loop  (cost=0.00..184831.04 rows=2448 width=13) (actual time=6.233..149.189 rows=273 loops=1)',),
 ("              Join Filter: (st_distance(r.geom, p.geom) < '10'::double precision)",),
 ('              Rows Removed by Join Filter: 7070',),
 ('              ->  Seq Scan on ne_10m_rivers_lake_centerlines r  (cost=0.00..351.82 rows=1 width=2866) (actual time=0.543..1.347 rows=1 loops=1)',),
 ("                    Filter: (((name_zh)::text ~~ '%亚马逊%'::text) AND ((featurecla)::text = 'River'::text))",),
 ('                    Rows Removed by Filter: 1454',),
 ('              ->  Seq Scan on ne_10m_populated_places p  (cost=0.00..812.43 rows=7343 width=45) (actual time=0.009..3.062 rows=7343 loops=1)',),
 ('Planning Time: 1.893 ms',),
 ('Execution Time: 153.055 ms',)]

基于上述查询规划回答以下问题

#### 5.3 为上述三个关系创建GiST空间索引（3分）

In [21]:
%%sql
drop index if exists rivers_geom_idx;
drop index if exists cities_geom_idx;
drop index if exists urban_areas_geom_idx;
create index rivers_geom_idx on ne_10m_rivers_lake_centerlines using GIST(geom);
create index urban_areas_geom_idx on ne_10m_urban_areas using GIST(geom);
create index cities_geom_idx on ne_10m_populated_places using GIST(geom);

 * postgresql://postgres:***@localhost:5432/hw4
Done.
Done.
Done.
Done.
Done.
Done.


[]

通过pg_class关系查询索引存储所需的数据页数和行数，即Block数目和城市区域、河流与城市数目

In [22]:
%sql select relpages, reltuples from pg_class where relname = 'urban_areas_geom_idx'

 * postgresql://postgres:***@localhost:5432/hw4
1 rows affected.


[(86, 11878.0)]

In [23]:
%sql select relpages, reltuples from pg_class where relname = 'rivers_geom_idx'

 * postgresql://postgres:***@localhost:5432/hw4
1 rows affected.


[(9, 1455.0)]

In [24]:
%sql select relpages, reltuples from pg_class where relname = 'cities_geom_idx'

 * postgresql://postgres:***@localhost:5432/hw4
1 rows affected.


[(47, 7343.0)]

#### 5.4 再次执行5.1的查询语句（3分）

In [25]:
%%sql
explain analyze
select r.name,count(*) num from ne_10m_rivers_lake_centerlines r, ne_10m_urban_areas u 
where r.featurecla = 'River' and ST_Crosses(r.geom,u.geom) group by r.name order by num desc;

 * postgresql://postgres:***@localhost:5432/hw4
15 rows affected.


[('Sort  (cost=35604.24..35606.67 rows=974 width=16) (actual time=2376.822..2376.835 rows=490 loops=1)',),
 ('  Sort Key: (count(*)) DESC',),
 ('  Sort Method: quicksort  Memory: 52kB',),
 ('  ->  HashAggregate  (cost=35546.15..35555.89 rows=974 width=16) (actual time=2376.695..2376.752 rows=490 loops=1)',),
 ('        Group Key: r.name',),
 ('        ->  Nested Loop  (cost=0.15..34755.54 rows=158121 width=8) (actual time=0.502..2374.822 rows=1867 loops=1)',),
 ('              ->  Seq Scan on ne_10m_rivers_lake_centerlines r  (cost=0.00..348.19 rows=1202 width=2874) (actual time=0.008..0.915 rows=1202 loops=1)',),
 ("                    Filter: ((featurecla)::text = 'River'::text)",),
 ('                    Rows Removed by Filter: 253',),
 ('              ->  Index Scan using urban_areas_geom_idx on ne_10m_urban_areas u  (cost=0.15..28.62 rows=1 width=1601) (actual time=0.365..1.969 rows=2 loops=1202)',),
 ('                    Index Cond: (geom && r.geom)',),
 ('                    Filter: st_crosses(r.geom, geom)',),
 ('                    Rows Removed by Filter: 8',),
 ('Planning Time: 2.495 ms',),
 ('Execution Time: 2377.011 ms',)]

基于上述查询规划回答以下问题

#### 5.5 再次查询5.2，使用ST_DWithin实现（4分）

In [26]:
%%sql 
explain analyze 
select distinct p.gid, p.name from ne_10m_rivers_lake_centerlines r, ne_10m_populated_places p 
where r.featurecla = 'River' and r.name_zh like '%亚马逊%' and ST_DWithin(r.geom,p.geom,10);

 * postgresql://postgres:***@localhost:5432/hw4
14 rows affected.


[('Unique  (cost=387.36..387.87 rows=69 width=13) (actual time=6.064..6.102 rows=273 loops=1)',),
 ('  ->  Sort  (cost=387.36..387.53 rows=69 width=13) (actual time=6.063..6.071 rows=273 loops=1)',),
 ('        Sort Key: p.gid, p.name',),
 ('        Sort Method: quicksort  Memory: 39kB',),
 ('        ->  Nested Loop  (cost=0.40..385.25 rows=69 width=13) (actual time=0.285..6.003 rows=273 loops=1)',),
 ('              ->  Seq Scan on ne_10m_rivers_lake_centerlines r  (cost=0.00..351.82 rows=1 width=2866) (actual time=0.200..0.503 rows=1 loops=1)',),
 ("                    Filter: (((name_zh)::text ~~ '%亚马逊%'::text) AND ((featurecla)::text = 'River'::text))",),
 ('                    Rows Removed by Filter: 1454',),
 ('              ->  Index Scan using cities_geom_idx on ne_10m_populated_places p  (cost=0.40..33.41 rows=1 width=45) (actual time=0.057..5.450 rows=273 loops=1)',),
 ("                    Index Cond: (geom && st_expand(r.geom, '10'::double precision))",),
 ("                    Filter: st_dwithin(r.geom, geom, '10'::double precision)",),
 ('                    Rows Removed by Filter: 43',),
 ('Planning Time: 1.109 ms',),
 ('Execution Time: 6.295 ms',)]

基于上述查询规划回答以下问题

#### 5.6 设置查询选项，如enable_indexscan，再次执行5.5查询语句（2分）

In [29]:
%sql SET enable_indexscan = false;

 * postgresql://postgres:***@localhost:5432/hw4
Done.


[]

In [30]:
%%sql 
explain analyze
select p.gid, p.name from ne_10m_rivers_lake_centerlines r, ne_10m_populated_places p 
where r.featurecla = 'River' and r.name_zh like '%亚马逊%' and ST_DWithin(r.geom,p.geom,10);

 * postgresql://postgres:***@localhost:5432/hw4
12 rows affected.


[('Nested Loop  (cost=4.41..385.25 rows=69 width=13) (actual time=0.321..6.063 rows=273 loops=1)',),
 ('  ->  Seq Scan on ne_10m_rivers_lake_centerlines r  (cost=0.00..351.82 rows=1 width=2866) (actual time=0.199..0.499 rows=1 loops=1)',),
 ("        Filter: (((name_zh)::text ~~ '%亚马逊%'::text) AND ((featurecla)::text = 'River'::text))",),
 ('        Rows Removed by Filter: 1454',),
 ('  ->  Bitmap Heap Scan on ne_10m_populated_places p  (cost=4.41..33.42 rows=1 width=45) (actual time=0.116..5.534 rows=273 loops=1)',),
 ("        Filter: st_dwithin(r.geom, geom, '10'::double precision)",),
 ('        Rows Removed by Filter: 43',),
 ('        Heap Blocks: exact=97',),
 ('        ->  Bitmap Index Scan on cities_geom_idx  (cost=0.00..4.41 rows=1 width=0) (actual time=0.059..0.059 rows=316 loops=1)',),
 ("              Index Cond: (geom && st_expand(r.geom, '10'::double precision))",),
 ('Planning Time: 0.236 ms',),
 ('Execution Time: 6.106 ms',)]

基于上述查询规划，回答该查询是否利用了索引，哪些索引，如何用来加速判断？

In [31]:
%sql SET enable_bitmapscan = false;

 * postgresql://postgres:***@localhost:5432/hw4
Done.


[]

In [32]:
%%sql 
explain analyze
select p.gid, p.name from ne_10m_rivers_lake_centerlines r, ne_10m_populated_places p 
where r.featurecla = 'River' and r.name_zh like '%亚马逊%' and ST_DWithin(r.geom,p.geom,10);

 * postgresql://postgres:***@localhost:5432/hw4
9 rows affected.


[('Nested Loop  (cost=0.00..184812.69 rows=69 width=13) (actual time=6.005..141.528 rows=273 loops=1)',),
 ("  Join Filter: st_dwithin(r.geom, p.geom, '10'::double precision)",),
 ('  Rows Removed by Join Filter: 7070',),
 ('  ->  Seq Scan on ne_10m_rivers_lake_centerlines r  (cost=0.00..351.82 rows=1 width=2866) (actual time=0.240..0.593 rows=1 loops=1)',),
 ("        Filter: (((name_zh)::text ~~ '%亚马逊%'::text) AND ((featurecla)::text = 'River'::text))",),
 ('        Rows Removed by Filter: 1454',),
 ('  ->  Seq Scan on ne_10m_populated_places p  (cost=0.00..812.43 rows=7343 width=45) (actual time=0.007..1.189 rows=7343 loops=1)',),
 ('Planning Time: 0.229 ms',),
 ('Execution Time: 141.562 ms',)]

基于上述查询规划，回答该查询是否利用了索引，哪些索引，如何用来加速判断？

In [33]:
%%sql
SET enable_indexscan = true;
SET enable_bitmapscan = true;

 * postgresql://postgres:***@localhost:5432/hw4
Done.
Done.


[]

### 6. 查询规划选择 (附加题3分)

关系R(A, B), S(B, C), T(C, D)统计信息如下：<br/>
T(R) = $10^5$, T(S) = 6 \* $10^6$, T(T) = 5 \* $10^4$<br/>
B(R) = 100, B(S) = 3000, B(T) = 40000<br/>
V(R, A) = 5 \* $10^4$, V(R,B) = V(S, B) = 3 \* $10^3$, V(S, C) = V(T, C) = 2 \* $10^4$, V(T, D) = $10^4$<br/>
创建了非聚集索引R.A, R.B, S.C, T.D和聚集索引S.B, T.C。

6.1 查询结果行数估计<br/>
&ensp;&ensp;&ensp; Select \*<br/>
&ensp;&ensp;&ensp; From R, S, T<br/>
&ensp;&ensp;&ensp; Where R.A = 2432 and R.B = S.B and S.C = T.C and T.D = 1234;

6.2 物理查询规划1的cost = 2 + 2 \* 1 + 4000 \* 2 = 8004 (具体看Lecture 8)
<img width = 500 src = 'physicalplan1.png' />
6.3 物理查询规划2如下图所示，估计对应的cost
<img width = 500 src = 'physicalplan2.png' />

### 作业感想

收获:-)，疑惑:-|，吐槽:-(，...，你的反馈很重要